sympy/combinatorics/fp_groups.py

```diff
@@ -514,69 +514,93 @@ def elements(self):
     >>> T.order() == f.order()/H.order()
     True
 
     '''
-    def _get_map(fp_group, F):
+    def _get_pres(parent_group, F):
         from sympy.combinatorics.homomorphisms import homomorphism
         if isinstance(F, list):
-            K, T = fp_group.subgroup(F, homomorphism=True)
+            K, T = parent_group.subgroup(F, homomorphism=True)
             f_gens = T(K.generators)
             f_rels = T(K.relators)
         else:
             f_gens = F.generators
             f_rels = F.relators
         return f_gens, f_rels
 
+    def _check(F):
+        if ((isinstance(F, list) and not all(elem in free_group for elem in F))
+        or (isinstance(F, FpGroup) and not (F.free_group == free_group))):
+            raise ValueError("The elements groups must belong to"
+                                    "to the same free group as that of the parent group")
+
     # If no parent group is specified,
     # G is set to the parent `FpGroup`
-    if not fp_group:
-        fp_group = G
-
-    if isinstance(G, list) and not fp_group:
-        raise ValueError("The fp_group must be"
+    if not parent_group:
+        if isinstance(G, list):
+            raise ValueError("The parent_group must be"
                             "defined when the group is a list")
-    free_group = fp_group.free_group
+        parent_group = G
+
+    free_group = parent_group.free_group
 
-    if not isinstance(fp_group, FpGroup):
+    if not isinstance(parent_group, FpGroup):
         raise ValueError("The parent group must be an instance"
                                     "of FpGroup")
 
-    if ((isinstance(G, list) and not all(elem in free_group for elem in G))
-        or (isinstance(G, FpGroup) and not (G.free_group == free_group))):
-            raise ValueError("The elements groups must belong to"
-                                    "to the same free group as that of the parent group")
-    if ((isinstance(H, list) and not all(elem in free_group for elem in H))
-        or (isinstance(H, FpGroup) and not (H.free_group == free_group))):
-            raise ValueError("The elements groups must belong to"
-                                    "to the same free group as that of the parent group")
+    _check(G)
+    _check(H)
+
+    # If the generators of `G` is a proper subgroup of the
+    # generators of `parent_group` return the `FpGroup` on a new presentation.
+    if isinstance(G, list) and all(elem in parent_group.generators for elem in G):
+        G, T = parent_group.subgroup(G, homomorphism=True)
+        h_gens, h_rels = _get_pres(parent_group, H)
+        h_gens = T(h_gens)
+        h_rels = T(h_rels)
+        free_group = G.free_group
+        q_realtors = list(G.relators) + list(h_gens)
+        return FpGroup(free_group, q_realtors), T
 
-    g_gens, g_rels = _get_map(fp_group, G)
-    h_gens, h_rels = _get_map(fp_group, H)
+    g_gens, g_rels = _get_pres(parent_group, G)
+    h_gens, h_rels = _get_pres(parent_group, H)
```

Comment:
I think this could be laid out better. `h_gens, h_rels = ...` could go outside any if statements because you are going to compute them anyway. `G, T = parent_group.subgroup(G, homomorphism=True)` could be done whenever `G` is a list, same for `free_group = G.free_group`. And after that the rest is the same for both cases